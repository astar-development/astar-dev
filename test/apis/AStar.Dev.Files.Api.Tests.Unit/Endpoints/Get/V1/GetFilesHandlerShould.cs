using AStar.Dev.Test.Helpers.Minimal.Api;
using AStar.Dev.Utilities;
using DbContextHelpers.Fixtures;
using Microsoft.AspNetCore.Http;
using NSubstitute;

namespace AStar.Dev.Files.Api.Endpoints.Get.V1;

public class GetFilesHandlerShould : IClassFixture<FilesContextFixture>
{
    private readonly FilesContextFixture mockFilesContextFactory;
    private readonly TimeProvider        mockTimeProvider;

    public GetFilesHandlerShould(FilesContextFixture mockFilesContextFactory)
    {
        this.mockFilesContextFactory = mockFilesContextFactory;
        mockTimeProvider             = Substitute.For<TimeProvider>();
        mockTimeProvider.GetUtcNow().Returns(new DateTimeOffset(new (2025, 7, 13, 1, 2, 3, DateTimeKind.Utc)));
    }

    [Fact]
    public async Task ReturnBadRequestWhenDuplicatesAreRequested()
    {
        await using var mockContext = mockFilesContextFactory.Sut;
        var             sut         = new GetFilesHandler();

        var response = await sut.HandleAsync(new(), mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ShouldBe(Results.BadRequest());
    }

    [Theory]
    [InlineData(@"c:\temp", true,  0, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                              SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                  expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", false,  0, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters2(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", true,  5, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters3(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", false, 5, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters4(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }
}
