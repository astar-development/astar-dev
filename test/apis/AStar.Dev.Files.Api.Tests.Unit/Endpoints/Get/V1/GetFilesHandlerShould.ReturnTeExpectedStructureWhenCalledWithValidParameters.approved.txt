[
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000102_hd_638509609800324210_hd.jpg",
    "fileHandle": "c:\\temp\\AI-100000102_hd_638509609800324210_hd.jpg-20"
  },
  {
    "directoryName": "c:\\temp\\wwwroot - Copy\\AI",
    "fileName": "100000102_hd_638509609800324210_hd.jpg",
    "fileHandle": "c:\\temp\\wwwroot - Copy\\AI-100000102_hd_638509609800324210_hd.jpg-49"
  },
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000110_hd_638511176661199240_hd.jpg",
    "fileHandle": "c:\\temp\\AI-100000110_hd_638511176661199240_hd.jpg-21"
  },
  {
    "directoryName": "c:\\temp\\AI\\test",
    "fileName": "100000110_hd_638511176661199240_hd.jpg",
    "fileHandle": "c:\\temp\\AI\\test-100000110_hd_638511176661199240_hd.jpg-107"
  },
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000145_638523103212069150.jpg",
    "fileHandle": "c:\\temp\\AI-100000145_638523103212069150.jpg-22"
  },
  {
    "directoryName": "c:\\temp\\wwwroot - Copy\\AI",
    "fileName": "100000145_638523103212069150.jpg",
    "fileHandle": "c:\\temp\\wwwroot - Copy\\AI-100000145_638523103212069150.jpg-50"
  },
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000146_638523184773955790.jpg",
    "fileHandle": "c:\\temp\\AI-100000146_638523184773955790.jpg-23"
  },
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000150_638524300326765720.jpg",
    "fileHandle": "c:\\temp\\AI-100000150_638524300326765720.jpg-24"
  },
  {
    "directoryName": "c:\\temp\\wwwroot - Copy\\AI",
    "fileName": "100000150_638524300326765720.jpg",
    "fileHandle": "c:\\temp\\wwwroot - Copy\\AI-100000150_638524300326765720.jpg-51"
  },
  {
    "directoryName": "c:\\temp\\AI",
    "fileName": "100000205_638509236020965080.jpg",
    "fileHandle": "c:\\temp\\AI-100000205_638509236020965080.jpg-25"
  }
]